### PR TITLE
Notifications: increase efficiency and fix pagination offset issue

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Documentation Update

## Description

Right now we filter recently aggregated notifications at runtime, which worked before when there was no pagination, but now filtering is not a good idea combined with offset pagination because it throws off the calculations.

In my case on dev.to/notifications it breaks after the third page.

The filtering should be done in the SQL query before paginating, which is what we've switched to. I added a bunch of scopes to make the code clearer as well.

To avoid nesting tags I've also updated the HTML and the JS logic to insert it in the right place.

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/assets/javascripts/initializers/initNotifications.js [hunk 1]
```diff
@@ -171,17 +170,25 @@ function initPagination() {
...
+              const markup = html.trim();
+
+              if (markup) {
+                const container = document.getElementById('articles-list');
+
+                const newNotifications = document.createElement('div');
+                newNotifications.innerHTML = markup;
+
+                paginator.remove();
+                container.append(newNotifications);
```
Comment (rhymes) [inline app/assets/javascripts/initializers/initNotifications.js]: Right now we're inserting the HTML coming from the server (which includes the paginator) in the previous paginator, basically nesting HTML in itself, which is pretty ugly.

With this we append the new notifications at the end of the list and remove the current paginator which is not needed after rendering.

File: app/assets/stylesheets/notifications.scss [hunk 2]
```diff
@@ -410,5 +410,9 @@
 
   .load-more-wrapper {
     @include load-more;
+
+    button {
+      margin-top: 0;
```
Comment (rhymes) [inline app/assets/stylesheets/notifications.scss]: after having moved the "load more" button outside the notification list container we have too much margin

File: app/controllers/notifications_controller.rb [hunk 3]
```diff
@@ -21,9 +21,11 @@ def index
                      elsif params[:org_id].blank? && params[:filter].present?
                        filtered_notifications
                      else
-                       Notification.where(user_id: @user.id).order("notified_at DESC")
+                       @user.notifications
                      end
 
+    @notifications = @notifications.without_recently_aggregated_reactions_and_follows.order(notified_at: :desc)
```
Comment (rhymes) [inline app/controllers/notifications_controller.rb]: after selecting the correct set of notifications we filter out recently aggregated ones *before* ordering and pagination

File: app/views/notifications/_notifications_list.html.erb [hunk 4]
```diff
@@ -1,22 +1,9 @@
-<% num_notifications = @notifications.size %>
-
-<% @notifications.each_with_index do |notification, index| %>
-  <% next if notification.notified_at < 24.hours.ago && notification.aggregated? %>
```
Comment (rhymes) [inline app/views/notifications/_notifications_list.html.erb]: This has been moved in the SQL query up a level

File: app/views/notifications/_notifications_list.html.erb [hunk 5]
```diff
@@ -30,3 +17,10 @@
     If you see too many of these, please report it to <a href="mailto:yo@dev.to">yo@dev.to</a>!
   </div>
 <% end %>
+
+<% if @notifications.any? %>
```
Comment (rhymes) [inline app/views/notifications/_notifications_list.html.erb]: this was good where it was in the beginning, after the block of notifications :)

File: app/views/notifications/index.html.erb [hunk 6]
```diff
@@ -64,16 +64,16 @@
...
-      <% if @notifications.size > 7 %>
-        <div class="load-more-wrapper">
-          <button id="load-more-button" type="button">
-            Load More
-          </button>
-        </div>
-      <% end %>
     </div>
+
+    <% if @notifications.size > 7 %>
```
Comment (rhymes) [inline app/views/notifications/index.html.erb]: the load more is fine outside the container for notifications, because it isn't a notification itself

Comment (lightalloy) [reply app/views/notifications/index.html.erb]: Maybe it's worth to add a comment for this magic number.

Comment (rhymes) [reply app/views/notifications/index.html.erb]: Will do!

**Pull request merged**